x-pack/plugins/endpoint/server/routes/resolver/children.ts

```diff
@@ -41,46 +42,90 @@ export const validateChildren = {
     after: schema.maybe(schema.string()),
     limit: schema.number({ defaultValue: 10, min: 1, max: 100 }),
     legacyEndpointID: schema.maybe(schema.string()),
+    levels: schema.number({ defaultValue: 0, min: 0, max: 3 }),
   }),
 };
 
+async function potentiallyKillKibana(
+  client: IScopedClusterClient,
+  id: string,
+  legacyEndpointID: string | undefined,
+  pagination: PaginationParams,
+  levels: number
+): Promise<any> {
+  const childrenQuery = new ChildrenQuery(legacyEndpointID, pagination);
+  const lifecycleQuery = new LifecycleQuery(legacyEndpointID);
+
+  // Retrieve the related child process events for a given process
+  const { total, results: events, nextCursor } = await childrenQuery.search(client, id);
+  const childIDs = events.map(extractEntityID);
+
+  // Retrieve the lifecycle events for the child processes (e.g. started, terminated etc)
+  // this needs to fire after the above since we don't yet have the entity ids until we
+  // run the first query
+  const { results: lifecycleEvents } = await lifecycleQuery.search(client, ...childIDs);
+
+  // group all of the lifecycle events by the child process id
+  const lifecycleGroups = _.groupBy(lifecycleEvents, extractEntityID);
+  const children = await Promise.all(
+    Object.entries(lifecycleGroups).map(async ([entityID, group]) => {
+      // Because this is recursive and in a map call, this is going to issue a maximum of
+      // (2 * (m - 1)) ^ n queries where n is the level of children to display and m is the limit of
```

Comment:
Think that this might actually be something like `2 * (m ^ n)`... either way, it's bad.